src/main/java/org/skriptlang/skript/log/runtime/RuntimeErrorManager.java

```diff
@@ -0,0 +1,147 @@
+	 * Tracked consumers are maintained during refreshes.
+	 */
+	public static void refresh() {
+		long frameLength = SkriptConfig.runtimeErrorFrameDuration.value().getAs(Timespan.TimePeriod.TICK);
+
+		int errorLimit = SkriptConfig.runtimeErrorLimitTotal.value();
+		int errorLineLimit = SkriptConfig.runtimeErrorLimitLine.value();
+		int errorLineTimeout = SkriptConfig.runtimeErrorLimitLineTimeout.value();
+		int errorTimeoutLength = Math.max(SkriptConfig.runtimeErrorTimeoutDuration.value(), 1);
+		FrameLimit errorFrame = new FrameLimit(errorLimit, errorLineLimit, errorLineTimeout, errorTimeoutLength);
+
+		int warningLimit = SkriptConfig.runtimeWarningLimitTotal.value();
+		int warningLineLimit = SkriptConfig.runtimeWarningLimitLine.value();
+		int warningLineTimeout = SkriptConfig.runtimeWarningLimitLineTimeout.value();
+		int warningTimeoutLength = Math.max(SkriptConfig.runtimeWarningTimeoutDuration.value(), 1);
+		FrameLimit warningFrame = new FrameLimit(warningLimit, warningLineLimit, warningLineTimeout, warningTimeoutLength);
+
+		List<RuntimeErrorConsumer> oldConsumers = List.of();
+		if (instance != null) {
+			instance.close();
+			oldConsumers = instance.consumers;
+		}
+		instance = new RuntimeErrorManager(Math.max((int) frameLength, 1), errorFrame, warningFrame);
+		oldConsumers.forEach(consumer -> instance.addConsumer(consumer));
+	}
+
+	private final Frame errorFrame, warningFrame;
+	private final Task task;
+
+	private final List<RuntimeErrorConsumer> consumers = new ArrayList<>();
+
+	/**
+	 * Creates a new error manager, which also creates its own frames.
+	 * <br>
+	 * Must be closed when no longer being used.
+	 *
+	 * @param frameLength The length of a frame in ticks.
+	 * @param errorLimits The limits to the error frame.
+	 * @param warningLimits The limits to the warning frame.
+	 */
+	public RuntimeErrorManager(int frameLength, FrameLimit errorLimits, FrameLimit warningLimits) {
+		errorFrame = new Frame(errorLimits);
+		warningFrame = new Frame(warningLimits);
+		task = new Task(Skript.getInstance(), frameLength, frameLength, true) {
+			@Override
+			public void run() {
+				consumers.forEach(consumer -> consumer.printFrameOutput(errorFrame.getFrameOutput(), Level.SEVERE));
+				errorFrame.nextFrame();
+
+				consumers.forEach(consumer -> consumer.printFrameOutput(warningFrame.getFrameOutput(), Level.WARNING));
+				warningFrame.nextFrame();
+			}
+		};
+	}
+
+	/**
+	 * Emits a warning or error depending on severity. Errors are passed to their respective {@link Frame}s for processing.
+	 * @param error The error to emit.
+	 */
+	public void error(@NotNull RuntimeError error) {
+		// print if < limit
+		if (error.level() == Level.SEVERE && errorFrame.add(error)) {
+			consumers.forEach((consumer -> consumer.printError(error)));
+		}
+
+		if (error.level() == Level.WARNING && warningFrame.add(error)) {
+			consumers.forEach((consumer -> consumer.printError(error)));
+		}
```

Comment:
These could be combined, right?
```suggestion
		if ((error.level() == Level.SEVERE && errorFrame.add(error))
			|| (error.level() == Level.WARNING && warningFrame.add(error))) {
			consumers.forEach((consumer -> consumer.printError(error)));
		}
```